app/src/main/java/chat/rocket/android/contacts/worker/ContactSyncWorker.kt

```diff
@@ -0,0 +1,148 @@
+                    contactArrayList.forEachIndexed { index, contact ->
+                        run {
+                            if (strongHashes[index] in intersectionSet) {
+                                contact.setUsername(intersectionMap[strongHashes[index]])
+                            }
+                        }
+                    }
+                } finally {
+                    dbManager.processContacts(contactArrayList)
+                }
+
+            }
+            Timber.d("Contacts fetched in background.")
+        }
+        return Result.SUCCESS
+
+    }
+
+    private fun getContactList() {
+        val cr = applicationContext.contentResolver
+
+        val cur = cr.query(ContactsContract.Contacts.CONTENT_URI, null, null, null, null)
+
+        if ((cur?.count ?: 0) > 0) {
+            while (cur != null && cur.moveToNext()) {
+                val id = cur.getString(
+                        cur.getColumnIndex(ContactsContract.Contacts._ID))
+                val name = cur.getString(cur.getColumnIndex(
+                        ContactsContract.Contacts.DISPLAY_NAME))
+
+                if (cur.getInt(cur.getColumnIndex(ContactsContract.Contacts.HAS_PHONE_NUMBER)) > 0) {
+                    // Has phone numbers
+
+                    val pCur = cr.query(
+                            ContactsContract.CommonDataKinds.Phone.CONTENT_URI, null,
+                            ContactsContract.CommonDataKinds.Phone.CONTACT_ID + " = ?",
+                            arrayOf<String>(id), null)
+                    while (pCur!!.moveToNext()) {
+                        val phoneNo = pCur.getString(pCur.getColumnIndex(
+                                ContactsContract.CommonDataKinds.Phone.NUMBER))
+                        val contact = Contact()
+                        contact.setName(name)
+                        contact.setPhoneNumber(phoneNo)
+                        contactArrayList.add(contact)
+                    }
+                    pCur.close()
+                }
+
+                if (true) {
+                    // No check for having email address
+
+                    val eCur = cr.query(
+                            ContactsContract.CommonDataKinds.Email.CONTENT_URI, null,
+                            ContactsContract.CommonDataKinds.Email.CONTACT_ID + " = ?",
+                            arrayOf<String>(id), null)
+                    while (eCur!!.moveToNext()) {
+                        val emailID = eCur.getString(eCur.getColumnIndex(
+                                ContactsContract.CommonDataKinds.Email.DATA))
+                        val contact = Contact()
+                        contact.setName(name)
+                        contact.setEmailAddress(emailID)
+                        contactArrayList.add(contact)
+                    }
+                    eCur.close()
+                }
+            }
+        }
+        cur?.close()
+        contactArrayList.sortWith(Comparator { o1, o2 ->
```

Comment:
Is there a reason to perform this sort at this point in time? I mean, the contacts will be displayed out of the CONTACTS table, right? So is there any benefit to sorting this at this point?

Reply:
After this, the contacts are inserted into table in this sorted order, so that they can be fetched and rendered directly and there is no need to sort them while rendering.

Reply:
Is the CONTACTS table recreated every time? Or, does it get updated with just changed records after the first time?

Reply:
It gets recreated every time.

Reply:
I see. So it may get read from multiple times and written to 1) at startup, 2) when the refresh button is pressed, and possibly 3) when a change to contacts db on device is detected. Is that correct?

Reply:
It would be great if you could show the old one while a new contacts sync was in progress in the background. And while that is happening we can show the spinner..

![image](https://user-images.githubusercontent.com/4001792/49946922-f48dae00-febd-11e8-93a6-a7b5730c9447.png)

Reply:
Right now it is written just at startup. We can make it write when the refresh button is pressed. It doesn't support automated detection of changed in the contacts DB.
The old entries in the table are deleted right before inserting new entires (The new entries have been fetched and processed), if the user clicks the FAB before this happens, they will see the old data. After the UI functionality is built, we can add an observer for the worker, using which we can set the state of the spinner.